tests/test_misc.rs

```diff
@@ -80,3 +80,115 @@ fn test_big_endian() {
     // so we can see it in the CI logs
     assert_eq!(u16::to_be_bytes(1), [0, 1]);
 }
+
+#[test]
+fn test_new_encoder() {
+    let arr3 = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]; // 2 * 2 * 3
+    let arr4 = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15]; // 2 * 2 * 4
+
+    let enc = qoi::Encoder::new(&arr3, 2, 2).unwrap();
+    assert_eq!(enc.channels(), Channels::Rgb);
+
+    let enc = qoi::Encoder::new(&arr4, 2, 2).unwrap();
+    assert_eq!(enc.channels(), Channels::Rgba);
+
+    assert!(matches!(
+        qoi::Encoder::new(&arr3, 3, 3),
+        Err(Error::InvalidImageLength { size: 12, width: 3, height: 3 })
+    ));
+
+    assert!(matches!(
+        qoi::Encoder::new(&arr3, 1, 1),
+        Err(Error::InvalidImageLength { size: 12, width: 1, height: 1 })
+    ));
+
+    let enc = qoi::EncoderBuilder::new(&arr3, 2, 2)
+        .stride(2 * 3)
+        .raw_channels(RawChannels::Bgr)
+        .build()
+        .unwrap();
+    assert_eq!(enc.channels(), Channels::Rgb);
+    let qoi = enc.encode_to_vec().unwrap();
+    let (_header, res) = decode_to_vec(qoi).unwrap();
+    assert_eq!(res, [2, 1, 0, 5, 4, 3, 8, 7, 6, 11, 10, 9]);
```

Comment:
mb to reduce some copy/paste?
```suggestion
    let test = |src, stride, sch, ch, exp| -> Result<()> {
        let enc = qoi::EncoderBuilder::new(src, 2, 2)
            .stride(stride)
            .source_channels(sch)
            .build()?;
        assert_eq!(enc.channels(), ch);
        let qoi = enc.encode_to_vec()?;
        let (_, res) = decode_to_vec(qoi)?;
        assert_eq!(res, exp, "{stride} {sch:?} {ch:?}");
        Ok(())
    };
    
    test(&arr3, 2 * 3, SourceChannels::Bgr, Channels::Rgb, [2, 1...])?;
    ...
```

Reply:
why not, but it may be a macro instead as arguments have different types.